crates/red_knot_python_semantic/src/util/subscript.rs

```diff
@@ -1,83 +1,448 @@
+        // 'checked_neg' only fails for i32::MIN. We can not
+        // represent -i32::MIN as a i32, but we can represent
+        // it as a usize, since usize is at least 32 bits.
+        from_nonnegative_i32(i32::MAX) + 1
+    })
+}
+
+impl Nth {
+    fn from_index(index: i32) -> Self {
         if index >= 0 {
-            self.nth(usize::try_from(index).ok()?)
+            Nth::FromStart(from_nonnegative_i32(index))
         } else {
-            let nth_rev = usize::try_from(index.checked_neg()?).ok()?.checked_sub(1)?;
-            self.rev().nth(nth_rev)
+            Nth::FromEnd(from_negative_i32(index) - 1)
+        }
+    }
+
+    fn to_nonnegative_index(&self, len: usize) -> usize {
+        debug_assert!(len > 0);
+
+        match self {
+            Nth::FromStart(nth) => *nth,
+            Nth::FromEnd(nth_rev) => (len - 1).saturating_sub(*nth_rev),
+        }
+    }
+}
+
+impl<I, T> PyIndex for T
+where
+    T: DoubleEndedIterator<Item = I>,
+{
+    type Item = I;
+
+    fn py_index(&mut self, index: i32) -> Result<I, OutOfBoundsError> {
+        match Nth::from_index(index) {
+            Nth::FromStart(nth) => self.nth(nth).ok_or(OutOfBoundsError),
+            Nth::FromEnd(nth_rev) => self.rev().nth(nth_rev).ok_or(OutOfBoundsError),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub(crate) struct StepSizeZeroError;
+
+pub(crate) trait PySlice {
+    type Item;
+
+    fn py_slice<'a>(
+        &'a mut self,
+        start: Option<i32>,
+        stop: Option<i32>,
+        step: Option<i32>,
+    ) -> Result<Box<dyn Iterator<Item = Self::Item> + 'a>, StepSizeZeroError>
+    where
+        Self::Item: 'a;
+}
+
+impl<I, T> PySlice for T
+where
+    T: DoubleEndedIterator<Item = I> + ExactSizeIterator<Item = I> + Iterator<Item = I>,
+{
+    type Item = I;
+
+    fn py_slice<'a>(
+        &'a mut self,
+        start: Option<i32>,
+        stop: Option<i32>,
+        step_int: Option<i32>,
+    ) -> Result<Box<dyn Iterator<Item = I> + 'a>, StepSizeZeroError>
+    where
+        I: 'a,
+    {
+        let step_int = step_int.unwrap_or(1);
+        if step_int == 0 {
+            return Err(StepSizeZeroError);
+        }
+
+        let len = self.len();
+        if len == 0 {
+            return Ok(Box::new(std::iter::empty()));
+        }
+
+        let to_nonnegative_index = |index| Nth::from_index(index).to_nonnegative_index(len);
+
+        if step_int > 0 {
+            let step = from_nonnegative_i32(step_int);
+            let start = start.map(to_nonnegative_index).unwrap_or(0).clamp(0, len);
+            let stop = stop.map(to_nonnegative_index).unwrap_or(len).clamp(0, len);
+
+            let (skip, take, step) = match start.cmp(&stop) {
+                Ordering::Less => (start, stop - start, step),
+                Ordering::Equal | Ordering::Greater => (start, 0, step),
+            };
+
+            Ok(Box::new(self.skip(skip).take(take).step_by(step)))
```

Comment:
I think you could use `Itertools::iter::Either` to return one or the other iterator type without allocating

Reply:
Oh, cool — I didn't know about `itertools::Either`! Needed to adapt the zero-case a bit to also fit into the `Either::Left` side, but that's okay.